File: doc/haskell/Data/NonEmptyFootList.hs
{-# LANGUAGE TupleSections #-}

module Data.NonEmptyFootList where

import Data.ListX (descartesWith)
import Data.List (uncons)
import Control.Arrow (second)


type NonEmptyFootList a = ([a], a)

footMap f (as, a) = (map f as, f a)

footUncons :: NonEmptyFootList a -> Maybe (a, NonEmptyFootList a)
footUncons (lst, foot) = fmap (second (, foot)) $ uncons lst

footHead :: NonEmptyFootList a -> a
footHead (lst, foot) = maybe foot fst $ uncons lst

tripleToNonEmptyFootList :: (r, r, r) -> NonEmptyFootList r
tripleToNonEmptyFootList (a, b, c) = ([a, b], c)

quadrupleToNonEmptyFootList :: (r, r, r, r) -> NonEmptyFootList r
quadrupleToNonEmptyFootList (a, b, lambda, c) = ([a, b, lambda], c)

descartesFootPlus :: Num a => [NonEmptyFootList a] -> [NonEmptyFootList a] -> [NonEmptyFootList a]
descartesFootPlus = descartesWith neFootPlus

neFootPlus :: Num a => NonEmptyFootList a -> NonEmptyFootList a -> NonEmptyFootList a
neFootPlus (as, a) (bs, b) = (zipWith (+) as bs, a + b)

File: app/assets/math/haskell/app/NonEmptyFootList.js
function descartesFootPlus(fl1, fl2) {return descartesWith(footPlus, fl1, fl2);}

function footPlus(flA, flB) {return footLift2((a, b) => a + b, flA, flB);}

function footMap(f, [as, a]) {return [as.map(f), f(a)];}
function footLift2(f, [as, a], [bs, b]) {return [zipWith(f, as, bs), f(a, b)];}
function footUncons([as, a]) {return maybeMap(([a0, as_]) => [a0, [as_, a]], uncons(as));}


//tripleToNonEmptyFootList :: (r, r, r, r) -> NonEmptyFootList r
function tripleToNonEmptyFootList([a, b, rhs]) {return [[a, b], rhs];}

//quadrupleToNonEmptyFootList :: (r, r, r, r) -> NonEmptyFootList r
function quadrupleToNonEmptyFootList([a, b, lambda, c]) {return [[a, b, lambda], c];}
